### PR TITLE
DRAFT Version of Dataverse Workflow POC

### PR DESCRIPTION
**Pull request by ChrisMuller80**
Fixes DD-39 - POC handle licence field (WIP)

* See: https://drivenbydata.atlassian.net/browse/DD-39

#### Where should the reviewer start?

Note:   
I was not able to fix the issue that first an Ok() has to be sent before the resume request is done. 
Therefore the dataset is always locked when publishing. 

#### How should this be manually tested?

1. Check out this PR
2. Upload Workflow  
`curl http://localhost:8080/api/admin/workflows -X POST --data-binary @{absolute path}/dd-easy-worflows-poc/src/main/scala/nl/knaw/dans/dd/wf/workflows/license-workflow.json -H "Content-type: application/json"`
3. Set as default Pre-Publish Workflow:  
`curl http://localhost:8080/api/admin/workflows/default/PrePublishDataset -X PUT -d 1`
4. Enable the license metadatablock in Dataverse
5. Create a new dataset with a licence and an access level
6. Publish
7. Check terms tab.

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -1,9 +1,7 @@
 # configuration for authorized and required reviewers
-* @DANS-KNAW/easy-leads
 
 # code
 # no overrides
 
 # policies
-/.github/PULL_REQUEST_TEMPLATE.md @DANS-KNAW/easy-leads
-/.github/CODEOWNERS @DANS-KNAW/easy-leads
+
```
Comment (jo-pol) [inline .github/CODEOWNERS:9]: I suppose you should have replaced with @DANS-KNAW/dataversedans

The archetype also missed some lines for [PULL_REQUEST_TEMPLATE](https://github.com/DANS-KNAW/dd-easy-worflows-poc/blob/master/.github/PULL_REQUEST_TEMPLATE.md)
It should trigger the review process with notifications and get it on an overview like for [EASY](https://github.com/pulls?utf8=%E2%9C%93&q=is%3Aopen+is%3Apr+team%3ADANS-KNAW%2Feasy)



Its content should become something like
```
Fixes EASY-

* 
* 
* 

#### Where should the reviewer @DANS-KNAW/dataversedans start?

#### How should this be manually tested?

#### Related pull requests on github

repo                       | PR                | note
-------------------------- | ----------------- | ----
easy-                      | [PR#](PRlink)     | ..

```

**Review by jo-pol — CHANGES_REQUESTED** [review]

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        .header("accept", "application/json")
+        .header("X-Dataverse-key", configuration.apiToken)
+        .asString.body
+
+    datasetJson
+  }
+
+  def updateLicense(jsonString: String): String = {
+
+    var json = JsonMethods.parse(jsonString) \ "data"
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala:44]: please avoid `var`

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    val fields = json \ "metadataBlocks" \ "access-and-license" \ "fields"
+    val licenseFromMetadata = (fields(1) \ "value").extract[String]
+    val access = (fields(0) \ "value").extract[String]
+
+    if (access.equals("Open Access")) {
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala:50]: Please use the scala way to compare objects
```suggestion
    if (access == "Open Access") {
```

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      if (!licenseFromMetadata.equalsIgnoreCase("CC0-1.0")) {
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala:51]: idem
```suggestion
      if (!licenseFromMetadata.toUpperCase == "CC0-1.0") {
```

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+          case JField("termsOfUse", JString(_)) => ("termsOfUse", JString(licenseFromMetadata))
+        }
+      }
+    }
+    else {
+      json = json.transformField {
+        case JField("license", JString(_)) => ("license", JString("NONE"))
+        case JField("termsOfUse", JString(_)) => ("termsOfUse", JString("DANS LICENSE"))
+      }
+    }
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala:63]: I'm not familiar with `json.transformField` but can't we rewrite this into something like below? (there is still a logical expression to extract into a val) 

Please provide unit tests.

```
    val isOpen = access == "Open Access"
    val isCC0 = licenseFromMetadata.toUpperCase == "CC0-1.0"
    json = json.transformField {
      case JField("license", JString(_)) if (isOpen && !isCC0) || !isOpen => ("license", JString("NONE"))
      case JField("termsOfUse", JString(_)) if (isOpen && !isCC0) || !isOpen => ("termsOfUse", JString(licenseFromMetadata))
      case JField("termsOfUse", JString(_)) if !isOpen => ("termsOfUse", JString("DANS LICENSE"))
    }
```

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    write(json)
+  }
+
+  def updateMetadata(datasetIdentifier: String, invocationId: String, metadata: String): Unit = {
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala:67]:
```suggestion
  private def updateMetadata(datasetIdentifier: String, invocationId: String, metadata: String): Unit = {
```

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+      Http(s"${ configuration.baseUrl }/api/datasets/:persistentId/versions/:draft?persistentId=$datasetIdentifier")
+        .header("content-type", "application/json")
+        .header("accept", "application/json")
+        .header("X-Dataverse-key", configuration.apiToken)
+        .asString.body
+
+    datasetJson
+  }
+
+  def updateLicense(jsonString: String): String = {
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala:42]:
```suggestion
  private def updateLicense(jsonString: String): String = {
```

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+import scalaj.http.Http
+
+class DdEasyWorkflowsPocApp(configuration: Configuration) {
+  def doWorkflow(invocationId: String, datasetIdentifier: String): Unit = {
+    val json = getDatasetJson(datasetIdentifier)
+    val updatedJson = updateLicense(json)
+    updateMetadata(datasetIdentifier, invocationId, updatedJson)
+  }
+
+  def getDatasetJson(datasetIdentifier: String): String = {
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala:30]:
```suggestion
  private def getDatasetJson(datasetIdentifier: String): String = {
```

**Review by jo-pol — CHANGES_REQUESTED** [review]

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+    val datasetJson =
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala:32]: image

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+      Thread.sleep(4000)
+      Http(s"${ configuration.baseUrl }/api/workflows/$invocationId")
+        .postData("")
+        .header("content-type", "application/json")
+        .header("accept", "application/json")
+        .header("X-Dataverse-key", configuration.apiToken)
+        .asString.headers
+    })
+
+    resume.start()
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala:88]: inline here too

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala (hunk 7 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocApp.scala:42]: For your first scala unit tests you might want to keep this public. The unit tests then can become something like 
```
"updateLicense" should "succeed" in {
   new DdEasyWorkflowsPocApp(null).updateLicense("""..."".") shouldBe """..."""
}
it should "fail" in {
  the[...Exception] thrownBy ... shoulBe ...Eception(...)
}
```

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocServlet.scala [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package nl.knaw.dans.dd.wf
+
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+import org.json4s.native.JsonMethods
+import org.scalatra._
+
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocServlet.scala:21]:
```suggestion
import nl.knaw.dans.lib.error._
import scala.util.Try

```

File: src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocServlet.scala [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    contentType = "text/plain"
+    Ok(s"DD Easy Worflows Poc Service running ($version)")
+  }
+
+  post("/workflow") {
+    contentType = "application/json"
+    val requestBodyJson = JsonMethods.parse(request.body)
+    val invocationId = (requestBodyJson \ "invocationId").extract[String]
+    val datasetIdentifier = (requestBodyJson \ "globalId").extract[String]
+    app.doWorkflow(invocationId, datasetIdentifier)
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.dd.wf/DdEasyWorkflowsPocServlet.scala:35]: Minimal error handling
```suggestion
    (for {
      requestBodyJson <- Try(JsonMethods.parse(request.body))
      invocationId = (requestBodyJson \ "invocationId").extract[String]
      datasetIdentifier = (requestBodyJson \ "globalId").extract[String]
      _ <- Try(app.doWorkflow(invocationId, datasetIdentifier))
    } yield Ok())
      .doIfFailure {
        case e => logger.error(e.getMessage, e)
      }.getOrElse(InternalServerError())
```

Of course a json parse error should become a bad requests etc, but after all it's a proof of concept.

File: README.md [hunk 13]
```diff
@@ -1,56 +1,4 @@
...
-    mvn clean install
-
-If the `rpm` executable is found at `/usr/local/bin/rpm`, the build profile that includes the RPM 
-packaging will be activated. If `rpm` is available, but at a different path, then activate it by using
-Maven's `-P` switch: `mvn -Pprm install`.
-
-Alternatively, to build the tarball execute:
-
-    mvn clean install assembly:single
+For documentation see: https://dans-knaw.github.io/easy-module-archetype
```
Comment (jo-pol) [inline README.md:4]:
```suggestion
For documentation see: https://dans-knaw.github.io/dd-easy-worflows-poc
```

**Review by janvanmansum — COMMENTED** [review]
A POC does *not* have to have very clean code (it is not prohibited either,
of course), but is *should* be easy to run the POC. How to do that?

* Make sure it builds (now there are some unit test failing)
* Document *all* the steps for running the POC. 
  * How to start the Dataverse instance. You don't have to describe it again if
    it is already documented, of course. In this case Menko described that. But
    do refer where it is documented.
  * (I guess the API token should always be necessary, although to my surprise it
    wasn't. Maybe we should look into that.)
  * It didn't work on my machine, so I should have taken some additional 
    step.

**Comment by jo-pol** [comment]
I'm afraid a `k` is missing in the name of the project.

**Pull request merged**